src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -392,7 +392,7 @@ internal T get_Item<T>(int index)
             // ! Warning: "this" is an array, not an SZArrayHelper. See comments above
             // ! or you may introduce a security hole!
             T[] _this = Unsafe.As<T[]>(this);
-            if ((uint)index >= (uint)_this.Length)
+            if ((uint)index >= _this.Length)
```

Comment:
Is this going to end up using 64-bit comparison? That may be slow on 32-bit platforms.

Reply:
That would be bad 😢 

If we need to change back to the `uint`-cast for the length in order to have uint-comparison on both 64-bit and 32-bit, then the "goal" to remove the workarounds for the bound-checks is somewhat missed -- even if the bound-checks get elided.

Here in CoreLib we can do it, as we know about the tricks, and workarounds the JIT needs. But the standard-developer who wants fast .NET code may know to cast the index to uint, to avoid the extra >= 0 check, but then they produce bad code for 32-bit.

So ideally and hopefully the JIT can treat the length here (and on {ReadOnly}Span) as unsigned, to produce proper code on every platform and avoid the extra bound-check.

@EgorBo can you please comment on this?

Reply:
I think I've fixed it for x86 unintentionally in https://github.com/dotnet/runtime/pull/62864/files#diff-5b83397bbbdd17bb9457998b520fdaaa474d165390985b66f32371561b6d0bacR13996-R14001

![image](https://user-images.githubusercontent.com/523221/161381208-b620d245-eb02-4261-9be7-21106a82b20b.png)

while it is indeed an issue in .NET 6.0 as you can see from this screenshot:
![image](https://user-images.githubusercontent.com/523221/161381250-78e928dc-5aee-43c6-9fe3-88a83c6bf214.png)

it's not an issue in .NET 7.0. I think I recall my PR had a huge negative diff on x86 but I didn't pay attention to it as I was focused on x64 codegen

Reply:
> But the standard-developer who wants fast .NET code may know to cast the index to uint, to avoid the extra >= 0 check

If one is using the cast to `uint` trick, I think it is better for clarity to cast both sides to `uint` to make it clear what you wanted to happen.

Reply:
I agree. 
Although it's possible to use `(uint)index < array.Length` now it reads a bit strange. I'd prefer to have `(uint)` on both sides (at least now, having seen the new pattern).

So for this PR:
* revert to both sides having the `uint`-cast
* keep the `{array | span}.Length > constant` work
* keep the code work that removes bound-checks that were previously there

?

Reply:
Sounds good to me.

Reply:
>  I'd prefer to have (uint) on both sides (at least now, having seen the new pattern).

FWIW, I've also taken to just always using it on both sides as I can't help track of exactly in what situations the JIT comprehends, and it's simpler to just always be consistent and do the thing that has the best chance of being recognized.

Reply:
Updated, so that when `uint`-cast is done for the index, then the cast is for clarity on both sides.